cli/tools/pm/audit.rs

```diff
@@ -0,0 +1,538 @@
+use crate::sys::CliSys;
+
+pub async fn audit(
+  flags: Arc<Flags>,
+  audit_flags: AuditFlags,
+) -> Result<i32, AnyError> {
+  let factory = CliFactory::from_flags(flags);
+  let _cli_options = factory.cli_options()?;
+  let npm_resolver = factory.npm_resolver().await?;
+  let npm_resolver = npm_resolver.as_managed().unwrap();
+  let snapshot = npm_resolver.resolution().snapshot();
+
+  let sys = CliSys::default();
+  let npm_url = npm_registry_url(&sys);
+  let http_provider = HttpClientProvider::new(None, None);
+
+  // socket_dev::call_firewall_api(
+  //   &snapshot,
+  //   http_provider.get_or_create().unwrap(),
+  // )
+  // .await?;
+
+  npm::call_audits_api(
+    audit_flags,
+    npm_url,
+    &snapshot,
+    http_provider.get_or_create().unwrap(),
+  )
+  .await
+}
+
+mod npm {
+  use std::collections::HashMap;
+  use std::collections::HashSet;
+
+  use deno_core::anyhow::Context;
+  use deno_npm::NpmPackageId;
+
+  use super::*;
+
+  #[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
+  enum AdvisorySeverity {
+    Low,
+    Moderate,
+    High,
+    Critical,
+  }
+
+  impl AdvisorySeverity {
+    fn parse(str_: &str) -> Option<Self> {
+      match str_ {
+        "low" => Some(Self::Low),
+        "moderate" => Some(Self::Moderate),
+        "high" => Some(Self::High),
+        "critical" => Some(Self::Critical),
+        _ => None,
+      }
+    }
+  }
+
+  fn get_dependency_descriptors_for_deps(
+    seen: &mut HashSet<String>,
+    npm_resolution_snapshot: &NpmResolutionSnapshot,
+    package_id: &NpmPackageId,
+  ) -> HashMap<String, Box<DependencyDescriptor>> {
+    let resolution_package =
+      npm_resolution_snapshot.package_from_id(package_id).unwrap();
+    let mut deps_map =
+      HashMap::with_capacity(resolution_package.dependencies.len());
+    for dep in resolution_package.dependencies.iter() {
+      let seen_str =
+        format!("{}@{}", dep.0.to_string(), dep.1.nv.version.to_string());
+      if seen.contains(&seen_str) {
+        continue;
+      }
+      seen.insert(seen_str);
+
+      let dep_deps = get_dependency_descriptors_for_deps(
+        seen,
+        npm_resolution_snapshot,
+        dep.1,
+      );
+      deps_map.insert(
+        dep.0.to_string(),
+        Box::new(DependencyDescriptor {
+          version: dep.1.nv.version.to_string(),
+          // TODO(bartlomieju): not sure how to determine that from the snapshot
+          dev: false,
+          requires: dep_deps
+            .iter()
+            .map(|(k, v)| (k.to_string(), v.version.to_string()))
+            .collect(),
+          dependencies: dep_deps,
+        }),
+      );
+    }
+    deps_map
+  }
+
+  pub async fn call_audits_api_inner(
+    npm_url: Url,
+    client: HttpClient,
+    body: serde_json::Value,
+  ) -> Result<AuditResponse, AnyError> {
+    let url = npm_url.join("/-/npm/v1/security/audits").unwrap();
+    let future = client.post_json(url, &body)?.send().boxed_local();
```

Comment:
Perhaps this should use the http client that has a progress bar? I forget what it's called.